### ISSUE TITLE
[L] - Koji calls should be moved to a backend task worker

### ISSUE THREAD

**Issue by bowlofeggs** [fedora-infra/bodhi#3061]
Creating or editing updates with many builds takes a long time. Even 50 builds can take an extreme about of time, to the point of causing API timeouts. There is likely more than one cause for this (indices come to mind), but one item that comes to mind that would help us is to move Koji API calls from the web server to backend task workers instead.

This would help with another problem too: right now we make our Koji API calls in the middle of a database transaction. If the database transaction is rolled back after Koji write calls were made, nothing will roll back those Koji calls. I believe this happened yesterday with https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2019-9e9f81e581 - it seems that Bodhi untagged one of its builds from being a buildroot override (https://bodhi.fedoraproject.org/overrides/python3-cryptography-2.3-2.el7), but for some reason rolled back the transaction after doing this. This caused Bodhi to be in an inconsistent state because it recorded the override as being active even though it was untagged in Koji. If we instead waited to commit the transaction before queueing tasks to do backend work like this, we could ensure that we have better consistency. Of course, we'd need good error handling and state tracking around making sure that the Koji calls make it to Koji (otherwise we also end up being inconsistent), but that seems like an easier place to deal with it than what we do today. Plus, it'll make the web API faster.

I am putting this on the CI Gating project kanban board because we will have much larger updates with Rawhide Gating than we typically do today, and this will be a performance bottleneck if we don't solve it.

We need to finish #2851 before we can work on this issue.

**Comment by cverna** [comment]
Scope
--------

Move the write calls to koji in a celery tasks to benefit from celery retry features. We don't want to queue the celery task until the database transaction is successfully committed.

Notes : We currently have a mechanism in place for this for sending fedora-messaging. There might be possibility to reuse this. (Sqlalchemy trigger/hook)

**Comment by nphilipp** [comment]
## Story Time…

As a package maintainer creating or editing updates with many builds, …

1. I want that the Bodhi UI/API doesn't block on the related Koji API calls,

    so that I don't run into timeouts.

2. I want that related Koji API calls are only performed after the transaction has been committed to the database,

    to avoid inconsistencies between Bodhi and Koji.

## Acceptance Criteria

- [ ] A Celery task worker exists which performs the changes related to a newly created or edited update in Koji.
- [ ] Creating an update or editing it doesn't perform the Koji changes in a blocking fashion, but delegates them to the task worker after the transaction has been committed to the database.

## Notes

- Existing tests will have to be adapted accordingly, e.g. where we mock out koji calls now, we might have to mock out calls creating the respective backend tasks (and mock the koji calls in the tests for the task worker).

**Comment by nphilipp** [comment]
@fedora-infra/bodhi @StephenCoady @AdamSaleh 

Hmm, we should probably reflect error handling in the acceptance criteria. The question is how far should we go here?

At the time the task worker does things in Koji, the respective update has been created/edited (committed to the DB). Should we attempt to roll back the update (which may be impossible)? Alert the user? Bodhi maintainers?

As far as I understand it, we should lock the update until Koji has been updated, too, to avoid that a subsequent change is performed using inconsistent information, i.e. when Bodhi and Koji disagree on the status of the builds or similar.

**Comment by nphilipp** [comment]
@StephenCoady @cverna With #3798 merged, should we close this one out, or keep it around for handling errors in the backend task?

**Comment by mattiaverga** [comment]
I'm really scared about that commit having been merged for the reason pointed out by @nphilipp in the second to last comment... do we know how Bodhi will behave in case of a Koji outage with build tags going out of sync with updates?

Also I think that `tag_update_builds_task` could take advantage of koji.multicall() instead of calling koji for every build in an update.

**Comment by cverna** [comment]
> @StephenCoady @cverna With #3798 merged, should we close this one out, or keep it around for handling errors in the backend task?

We can either keep this one around and create a new issue just for the error handling. Either way works for me :)

**Comment by cverna** [comment]
> I'm really scared about that commit having been merged for the reason pointed out by @nphilipp in the second to last comment... do we know how Bodhi will behave in case of a Koji outage with build tags going out of sync with updates?
> 

Yeah that's a possible risk, but that should not be too bad. Updates will be created but the builds might not be tagged properly. That will be easy to track down and fix. In the end the koji tag is the source of truth for a build being available in a repo (updates-testing or updates) so in my opinion this is an acceptable risk.

> Also I think that `tag_update_builds_task` could take advantage of koji.multicall() instead of calling koji for every build in an update.

Sounds good, do you want to create an issue or PR for that ?

**Comment by mattiaverga** [comment]
Closing as completed
